compiler/rustc_lint/src/gpukernel_abi.rs

```diff
@@ -0,0 +1,177 @@
+    ///  --> t.rs:1:1
+    ///   |
+    /// 1 | extern "gpu-kernel" fn kernel() {}
+    ///   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+    ///   |
+    ///   = help: use `unsafe(no_mangle)` or `unsafe(export_name = "<name>")`
+    ///   = note: mangled names make it hard to find the kernel, this is usually not intended
+    ///   = note: `#[warn(missing_gpu_kernel_export_name)]` on by default
+    /// ```
+    ///
+    /// ### Explanation
+    ///
+    /// `gpu-kernel` functions are usually searched by name in the compiled file.
+    /// A mangled name is usually unintentional as it would need to be searched by the mangled name.
+    ///
+    /// To use an unmangled name for the kernel, either `no_mangle` or `export_name` can be used.
+    /// ```rust,ignore (fails on non-GPU targets)
+    /// // Can be found by the name "kernel"
+    /// #[unsafe(no_mangle)]
+    /// extern "gpu-kernel" fn kernel() { }
+    ///
+    /// // Can be found by the name "new_name"
+    /// #[unsafe(export_name = "new_name")]
+    /// extern "gpu-kernel" fn other_kernel() { }
+    /// ```
+    MISSING_GPU_KERNEL_EXPORT_NAME,
+    Warn,
+    "mangled gpu-kernel function"
+}
+
+declare_lint_pass!(ImproperGpuKernelLint => [
+    IMPROPER_GPU_KERNEL_ARG,
+    MISSING_GPU_KERNEL_EXPORT_NAME,
+]);
+
+/// `ImproperGpuKernelLint` checks `gpu-kernel` function definitions:
+///
+/// - `extern "gpu-kernel" fn` arguments should be primitive types.
+/// - `extern "gpu-kernel" fn` should have an unmangled name.
+impl<'tcx> LateLintPass<'tcx> for ImproperGpuKernelLint {
+    fn check_fn(
+        &mut self,
+        cx: &LateContext<'tcx>,
+        kind: hir::intravisit::FnKind<'tcx>,
+        decl: &'tcx hir::FnDecl<'_>,
+        _: &'tcx hir::Body<'_>,
+        span: Span,
+        id: LocalDefId,
+    ) {
+        use hir::intravisit::FnKind;
+
+        let abi = match kind {
+            FnKind::ItemFn(_, _, header, ..) => header.abi,
+            FnKind::Method(_, sig, ..) => sig.header.abi,
+            _ => return,
+        };
+
+        if abi != ExternAbi::GpuKernel {
+            return;
+        }
+
+        let sig = cx.tcx.fn_sig(id).instantiate_identity();
+        let sig = cx.tcx.instantiate_bound_regions_with_erased(sig);
+
+        for (input_ty, input_hir) in iter::zip(sig.inputs(), decl.inputs) {
+            let is_valid_arg = match input_ty.kind() {
+                ty::Bool
+                | ty::Char
+                | ty::Int(_)
+                | ty::Uint(_)
+                | ty::Float(_)
+                | ty::RawPtr(_, _) => true,
+
+                ty::Adt(_, _)
+                | ty::Alias(_, _)
+                | ty::Array(_, _)
+                | ty::Bound(_, _)
+                | ty::Closure(_, _)
+                | ty::Coroutine(_, _)
+                | ty::CoroutineClosure(_, _)
+                | ty::CoroutineWitness(..)
+                | ty::Dynamic(_, _)
+                | ty::Error(_)
```

Comment:
We should cease inspecting the type or just `continue` if the type is erroneous.